components/specialized-parser/src/runner.py

```diff
@@ -0,0 +1,412 @@
+from google.cloud.exceptions import InternalServerError
+from sqlalchemy.engine import Engine
+
+FilenamesPair = namedtuple("FilenamesPair", "original_filename txt_filename")
+
+
+PROCESSED_DOCUMENTS_TABLE_NAME = "eks.processed_documents"
+
+
+@dataclass
+class ProcessedDocument:
+    id: str
+    original_filename: str
+    results_file: str
+    run_id: str
+    entities: str
+
+
+USER_AGENT = "cloud-solutions/eks-docai-v1"
+
+
+class SpecializedParserJobRunner:
+    def __init__(
+        self,
+        job_config: JobConfig,
+        processor_config: ProcessorConfig,
+        alloydb_config: AlloyDBConfig,
+        bigquery_config: BigQueryConfig,
+    ):
+        self.job_config = job_config
+        self.processor_config = processor_config
+        self.alloydb_config = alloydb_config
+        self.bigquery_config = bigquery_config
+
+        self.alloydb_connection_pool = self.create_connection_pool(alloydb_config)
+        self.storage_client = storage.Client(
+            client_info=ClientInfo(user_agent=USER_AGENT)
+        )
+        self.bq_client = bigquery.Client(client_info=ClientInfo(user_agent=USER_AGENT))
+
+    def run(self):
+        print("Verifying AlloyDB output table")
+        self.verify_alloydb_table()
+        print("Starting Batch Processor operation")
+        batch_operation = self.call_batch_processor()
+        print("Waiting for Batch operation to finish")
+        individual_process_statuses = self.wait_for_completion_and_verify_success(
+            batch_operation
+        )
+        print(f"Parsing results from {self.job_config.gcs_output_uri}")
+        parsed_results, filename_pairs = self.read_and_parse_batch_results(
+            individual_process_statuses,
+        )
+
+        print("Writing metadata to bigquery")
+        self.write_metadata_to_bigquery(filename_pairs)
+        if not parsed_results:
+            print("No parsed results from processor - only metadata")
+        else:
+            print("Writing results to GCS")
+            bucket_name, csv_blob_name = self.write_results_to_gcs(parsed_results)
+            print("Writing results to AlloyDB")
+            self.write_results_to_alloydb_with_inserts(parsed_results)
+            print("Writing results to BigQuery")
+            self.write_results_to_bigquery(bucket_name, csv_blob_name)
+        print("Done")
+
+    @staticmethod
+    def create_connection_pool(
+        alloydb_config: AlloyDBConfig, refresh_strategy: str = "lazy"
+    ) -> Engine:
+        connector = Connector(refresh_strategy=refresh_strategy)
+
+        def getconn() -> (
+            pg8000.dbapi.Connection
+        ):  # pyright: ignore [reportAttributeAccessIssue]
+            conn = connector.connect(
+                instance_uri=alloydb_config.primary_instance,
+                driver="pg8000",
+                db=alloydb_config.database,
+                enable_iam_auth=True,
+                user=os.environ["ALLOYDB_USER"],
```

Comment:
can we take the user as input parameter instead, maybe saved as instance variable, to avoid loading of the env from os here

Reply:
This one is set from terraform - the "user" is always the service account user, which is created and managed by terraform, so setting it as a default environment variable makes sense to me. if you still disagree, let's talk.

Reply:
fixed